src/test/util/blocktools.cpp

```diff
@@ -0,0 +1,116 @@
+// Copyright (c) 2019 The Unit-e developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <test/util/blocktools.h>
+
+#include <boost/test/unit_test.hpp>
+
+namespace blocktools {
+
+CBlockIndex *BlockIndexFake::MakeBlockIndex(const uint256 &hash, CBlockIndex *const prev) {
+  auto result = block_indexes.emplace(hash, CBlockIndex());
+  CBlockIndex *const index = &result.first->second;
+  index->phashBlock = &result.first->first;
+  index->nHeight = prev ? prev->nHeight + 1 : 0;
+  index->pprev = prev;
+  return index;
+}
+
+CBlockIndex *BlockIndexFake::Generate(const std::size_t count, const CBlockIndex *starting_point) {
+
+  std::size_t height = 0;
+  CBlockIndex *const starting_index = [&]() {
+    if (starting_point == nullptr) {
+      ++height;
+      const uint256 hash = GetRandHash();
+      return MakeBlockIndex(hash, nullptr);
+    }
+    auto it = block_indexes.find(starting_point->GetBlockHash());
+    BOOST_REQUIRE_MESSAGE(
+        it != block_indexes.end(),
+        "starting_point not known by this instance of BlockIndexFake.");
+    return &it->second;
+  }();
+  BOOST_REQUIRE(starting_index);
+  BOOST_REQUIRE(starting_index->phashBlock);
+
+  CBlockIndex *current_index = starting_index;
+  for (; height < count; ++height) {
+    const uint256 hash = GetRandHash();
+    current_index = MakeBlockIndex(hash, current_index);
+    BOOST_REQUIRE(current_index);
+    BOOST_REQUIRE(current_index->phashBlock);
+    BOOST_REQUIRE(current_index->pprev);
+  }
+
+  return current_index;
+}
+
+std::vector<CBlockIndex *> BlockIndexFake::GetChain(const CBlockIndex *tip) {
+  BOOST_REQUIRE(tip != nullptr);
+  auto it = block_indexes.find(tip->GetBlockHash());
+  BOOST_REQUIRE_MESSAGE(
+      it != block_indexes.end(),
+      "tip not known by this instance of BlockIndexFake.");
+  std::vector<CBlockIndex *> result;
+  CBlockIndex *found_tip = &it->second;
```

Comment:
I think, all lines above that could be replaced by

```C++
BOOST_REQUIRE_MESSAGE(
    block_indexes.count(tip->GetBlockHash()) != 0
    "tip not known by this instance of BlockIndexFake.");
```

and then we don't need `found_tip` at all.

Reply:
`tip` is a `const CBlockIndex*` and I wanted to keep it this way such that this function can be called also with results from `ActiveChain::GetTip()` (for instance).

`found_tip` is a `CBlockIndex*` which I also wanted to keep like this, such that tests can use the result to further alter the layout of the chain if need be.